source/adaptive_load/metrics_plugin_impl.cc

```diff
@@ -0,0 +1,221 @@
+    if (counter.name() == counter_name) {
+      return counter.value();
+    }
+  }
+  return absl::InternalError(
+      absl::StrCat("Counter '", counter_name, "' not found in Result proto."));
+}
+
+/**
+ * Finds a Statistic proto with the given id within a Result proto.
+ *
+ * @param result A Result proto taken from a Nighthawk Output proto.
+ * @param statistic_id The name of the desired Statistic, e.g.
+ * "benchmark_http_client.request_to_response".
+ *
+ * @return StatusOr<Statistic> The Statistic with the specified id if found, otherwise an error
+ * Status.
+ */
+absl::StatusOr<nighthawk::client::Statistic> GetStatistic(const nighthawk::client::Result& result,
+                                                          absl::string_view statistic_id) {
+  for (const nighthawk::client::Statistic& statistic : result.statistics()) {
+    if (statistic.id() == statistic_id) {
+      return statistic;
+    }
+  }
+  return absl::InternalError(
+      absl::StrCat("Statistic '", statistic_id, "' not found in Result proto."));
+}
+
+/**
+ * Extracts a value from a StatusOr into a variable if OK; otherwise appends the error message to a
+ * vector. static_casts the value to fit the variable. Ignores Status code values. For use in a
+ * constructor that stores a vector of errors during initialization.
+ *
+ * @param status_or The StatusOr to extract. The value must be static_castable to the type of
+ * |value|.
+ * @param value The variable where the value should be extracted. Only written if the status is OK.
+ * @param errors A place to accumulate error messages. If the status is not OK, its message is
+ * appended here.
+ *
+ * @return bool True if the status was OK.
+ */
+template <typename T, typename U>
+bool ExtractValueOrRecordError(absl::StatusOr<T> status_or, U& value,
+                               std::vector<std::string>& errors) {
+  if (status_or.ok()) {
+    value = static_cast<U>(status_or.value());
+    return true;
+  } else {
+    errors.emplace_back(status_or.status().message());
+    return false;
+  }
+}
+
+/**
+ * Extracts counters from a Nighthawk Service Output proto and computes metrics from them, storing
+ * the metrics in a map.
+ *
+ * @param nighthawk_output An Output proto returned from Nighthawk Service.
+ * @param metric_from_name A map to write computed metrics under various names.
+ * @param errors A place to accumulate error messages.
+ */
+void ExtractCounters(const nighthawk::client::Output& nighthawk_output,
+                     absl::flat_hash_map<std::string, double>& metric_from_name,
+                     std::vector<std::string>& errors) {
+  nighthawk::client::Result global_result;
+  if (!ExtractValueOrRecordError(GetResult(nighthawk_output, "global"), global_result, errors)) {
+    return;
+  }
+  const int64_t actual_duration_seconds =
+      Envoy::Protobuf::util::TimeUtil::DurationToSeconds(global_result.execution_duration());
```

Comment:
Have you considered aliasing these long statements (this one and the one to Nanoseconds) via a using statement in the anonymous namespace?

Reply:
Done